app/gui2/src/components/CodeEditor.vue

```diff
@@ -0,0 +1,66 @@
+<script setup lang="ts">
+import { ref, watchEffect } from 'vue'
+import { useWindowEvent } from '@/util/events'
+
+const emit = defineEmits<{
+  codeUpdate: [code: string]
+}>()
+
+const content = ref('main = \n    2 + 2')
+const shown = ref(false)
+
+watchEffect(() => {
+  emit('codeUpdate', content.value)
+})
+
+const textArea = ref<InstanceType<typeof HTMLTextAreaElement> | null>(null)
```

Comment:
```suggestion
const textArea = ref<HTMLElement>()
```

By default, a `ref` with a specified `T` type and no value actually returns a `Ref<T | undefined>`, so it's a convenient and correct notation. You can be more specific with the type of the element if you rely on it, but usually that's not necessary and `HTMLElement` is fine.

The `null`/`undefined` mix is unfortunate in JS, but my usual approach to it is use whatever is convenient at given place, and make sure that the code works the same with both. See `util/opt.ts` module for more details about that.